File: backend/app/api/v1/generation.py
import json

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sse_starlette.sse import EventSourceResponse

from app.api.schemas.schemas import GenerateRequest
from app.dependencies import get_db, get_project
from app.models.models import Project
from app.core.llm.output_parsers import extract_latex
from app.services.generation_service import (
    generate_latex_from_documents,
    generate_latex_pipeline,
)
from app.services import project_service

router = APIRouter(tags=["generation"])


@router.post("/projects/{project_id}/generate")
async def generate_latex(
    data: GenerateRequest,
    project: Project = Depends(get_project),
    db: AsyncSession = Depends(get_db),
):
    """Generate LaTeX with pipeline (progress events for multi-doc, simple for few docs)."""
    template_id = data.template_id or "academic_paper"

    async def event_stream():
        full_content = ""
        try:
            async for event in generate_latex_pipeline(
                db, project.id, template_id, data.document_ids
            ):
                evt_type = event.get("event", "")

                if evt_type == "stage":
                    yield {
                        "event": "stage",
                        "data": json.dumps({
                            "stage": event.get("stage"),
                            "message": event.get("message"),
                            "progress": event.get("progress", 0),
                            "detail": event.get("detail", ""),
                        }),
                    }

                elif evt_type == "outline":
                    yield {
                        "event": "outline",
                        "data": json.dumps({
                            "message": event.get("message"),
                            "outline": event.get("outline"),
                        }),
                    }

                elif evt_type == "chunk":
                    content = event.get("content", "")
                    full_content += content
                    yield {
                        "event": "chunk",
                        "data": json.dumps({"content": content}),
                    }

                elif evt_type == "done":
                    cleaned = event.get("content", extract_latex(full_content))
                    await project_service.update_project(
                        db, project, latex_content=cleaned, template_id=template_id
                    )
                    yield {
                        "event": "done",
                        "data": json.dumps({
                            "content": cleaned,
                            "message": event.get("message", ""),
                        }),
                    }

                elif evt_type == "error":
                    yield {
                        "event": "error",
                        "data": json.dumps({"error": event.get("message", "Unknown error")}),
                    }

        except Exception as e:
            yield {"event": "error", "data": json.dumps({"error": str(e)})}

    return EventSourceResponse(event_stream())

File: frontend/src/api/generation.ts
import { fetchSSE, parseSSEStream } from '../utils/sse';

export interface GenerateEvent {
  type: 'chunk' | 'done' | 'error' | 'stage' | 'outline';
  content: string;
  stage?: string;
  message?: string;
  progress?: number;
  detail?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  outline?: any;
}

export async function* generateLatex(
  projectId: string,
  options?: { template_id?: string },
  signal?: AbortSignal
): AsyncGenerator<GenerateEvent> {
  const response = await fetchSSE(
    `/api/v1/projects/${projectId}/generate`,
    { template_id: options?.template_id || '' },
    signal
  );

  for await (const raw of parseSSEStream(response)) {
    if (raw.data === '[DONE]') return;
    try {
      const parsed = JSON.parse(raw.data);
      if (raw.event === 'error') {
        yield { type: 'error', content: parsed.error || '' };
      } else if (raw.event === 'done') {
        yield {
          type: 'done',
          content: parsed.content || '',
          message: parsed.message,
        };
      } else if (raw.event === 'stage') {
        yield {
          type: 'stage',
          content: '',
          stage: parsed.stage,
          message: parsed.message,
          progress: parsed.progress,
          detail: parsed.detail,
        };
      } else if (raw.event === 'outline') {
        yield {
          type: 'outline',
          content: '',
          message: parsed.message,
          outline: parsed.outline,
        };
      } else {
        yield { type: 'chunk', content: parsed.content || '' };
      }
    } catch {
      // not JSON, skip
    }
  }
}
